### DISCUSSION TITLE
Can't use dataclasses as request body and response_model, even in different path operation functions

### DISCUSSION THREAD

**Discussion by mrchi** [fastapi/fastapi#8772] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from dataclasses import dataclass, field
from typing import List, Optional

from fastapi import FastAPI


@dataclass
class Item:
    name: str
    price: float
    tags: List[str] = field(default_factory=list)
    description: Optional[str] = None
    tax: Optional[float] = None


app = FastAPI()


@app.post("/items/next", response_model=Item)
async def read_next_item(item: Item):
    return {}
```

### Description

- Visit swagger docs `/docs`
- Get error `Failed to load API definition.`

```
Traceback (most recent call last):
  File "/.../.venv/lib/python3.9/site-packages/uvicorn/protocols/http/httptools_impl.py", line 398, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/.../.venv/lib/python3.9/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "/.../.venv/lib/python3.9/site-packages/fastapi/applications.py", line 199, in __call__
    await super().__call__(scope, receive, send)
  File "/.../.venv/lib/python3.9/site-packages/starlette/applications.py", line 112, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/.../.venv/lib/python3.9/site-packages/starlette/middleware/errors.py", line 181, in __call__
    raise exc from None
  File "/.../.venv/lib/python3.9/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "/.../.venv/lib/python3.9/site-packages/starlette/exceptions.py", line 82, in __call__
    raise exc from None
  File "/.../.venv/lib/python3.9/site-packages/starlette/exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "/.../.venv/lib/python3.9/site-packages/starlette/routing.py", line 580, in __call__
    await route.handle(scope, receive, send)
  File "/.../.venv/lib/python3.9/site-packages/starlette/routing.py", line 241, in handle
    await self.app(scope, receive, send)
  File "/.../.venv/lib/python3.9/site-packages/starlette/routing.py", line 52, in app
    response = await func(request)
  File "/.../.venv/lib/python3.9/site-packages/fastapi/applications.py", line 152, in openapi
    return JSONResponse(self.openapi())
  File "/.../.venv/lib/python3.9/site-packages/fastapi/applications.py", line 130, in openapi
    self.openapi_schema = get_openapi(
  File "/.../.venv/lib/python3.9/site-packages/fastapi/openapi/utils.py", line 376, in get_openapi
    definitions = get_model_definitions(
  File "/.../.venv/lib/python3.9/site-packages/fastapi/utils.py", line 28, in get_model_definitions
    model_name = model_name_map[model]
KeyError: <class 'pydantic.dataclasses.Item'>
```

### Operating System

macOS

### FastAPI Version

0.67.0

### Python Version

3.9.5

**Comment by mrchi** [comment]
That's interesting. I make futher tests, and found that it's not working even if I use dataclasses as request body and response model in different path operation functions.

Like this:

```python
from dataclasses import dataclass, field
from typing import List, Optional

from fastapi import FastAPI


@dataclass
class Item:
    name: str
    price: float
    tags: List[str] = field(default_factory=list)
    description: Optional[str] = None
    tax: Optional[float] = None


app = FastAPI()


@app.post("/items/one", response_model=Item)
async def one():
    return {}


@app.post("/items/two")
async def two(item: Item):
    return {}
```

Such a confused feature.

**Comment by whiskeyriver** [comment]
Have you tried with a pydantic dataclass?

https://pydantic-docs.helpmanual.io/usage/dataclasses/

**Comment by ringsaturn** [comment]
> Have you tried with a pydantic dataclass?
> 
> [pydantic-docs.helpmanual.io/usage/dataclasses](https://pydantic-docs.helpmanual.io/usage/dataclasses/)

Pydantic's BaseModel is well supported by FastAPI but our team needs a naive dataclass support for pydantic's dataclass has many BaseModel design patterns and needs many hacking to work with existing codes.

For example, if we want to support Mongo's `_id` field in Pydantic's BaseModel or dataclass, we need to hack an upper level dataclass like this ( Refer https://github.com/samuelcolvin/pydantic/issues/2277#issuecomment-764010272 ):

```python
class OID(str):
    @classmethod
    def __get_validators__(cls):
        yield cls.validate

    @classmethod
    def validate(cls, v):
        try:
            return ObjectId(str(v))
        except InvalidId:
            raise ValueError("Not a valid ObjectId")

class MongoModel(BaseModel):
    id: Optional[Union[str, OID]] = Field(alias="_id")

    class Config(BaseConfig):
        fields = {"id": "_id"}
        allow_population_by_field_name = True
        json_encoders = {
            ObjectId: lambda oid: str(oid),
        }
```

From personal options, this code looks very ugly and not pythonic.

**Comment by Mause** [comment]
@ringsaturn you misunderstand, pydantic also has support for dataclasses as well as `BaseModel`

**Comment by ringsaturn** [comment]
> @ringsaturn you misunderstand, pydantic also has support for dataclasses as well as `BaseModel`

Pydantic's dataclass has unnecessary limits compared with native dataclass.

```py
from typing import Any
from pydantic.dataclasses import dataclass as pydantic_dataclass
from dataclasses import dataclass


@dataclass
class Bar:
    _id: Any


print(Bar(_id=1))


@pydantic_dataclass
class Foo:
    _id: Any


print(Foo(_id=1))
```

Output:

```
Bar(_id=1)
app2.py:14: RuntimeWarning: fields may not start with an underscore, ignoring "_id"
  class Foo:
Traceback (most recent call last):
  File "app2.py", line 17, in <module>
    print(Foo(_id=1))
  File "xxxxxx/lib/python3.8/dataclasses.py", line 368, in wrapper
    result = user_function(self)
  File "<string>", line 3, in __repr__
AttributeError: 'Foo' object has no attribute '_id'
```

**Comment by whiskeyriver** [comment]
@ringsaturn 

I think I see what happens here. FastAPI will convert your plain dataclass to a Pydantic dataclass, and then extract the pydantic model at `new_dataclass.__pydantic_model__` in order to generate the openapi schema.  This seems natural because Pydantic will automatically convert your stdlib dataclass to a pydantic dataclass when its added to a BaseModel subclass. 

I didn't go through the whole exercise to trace it but the gist of it looks to be:
* It's seen in the route decorator, converted, and stored by its generated class name: 'pydantic.dataclasses.Item'
* It's seen in the function signature, converted, and updates the model map by its identical generated class name

Now model_map has only 1 entry for pydantic.dataclasses.Item, but each conversion created a different Pydantic object with a different hash, so when the first Item instance is sought in the model_map it's not found and you get the error.

```python
from dataclasses import dataclass, field
from typing import List, Optional

from pydantic.dataclasses import dataclass as pdc

@dataclass
class Test1:
    name: str
    price: float
    tags: List[str] = field(default_factory=list)
    description: Optional[str] = None
    tax: Optional[float] = None


if __name__ == "__main__":
    for _ in range(5):
        new_dc = pdc(Test1)
        print(hash(new_dc.__pydantic_model__))
```

```shell
$ python ./pdc.py
5878314349139
5878314350052
5878314350410
5878314351174
5878314351890
```

I would bet this isn't an easy fix. The generated class name of the pydantic model is very generic, and might collide with a different dataclass named Item in another file as well.

Was your Pydantic model example the one from https://www.mongodb.com/developer/quickstart/python-quickstart-fastapi/ ? It's pretty much the same thing as I do in my projects for SqlAlchemy models. I create Pydantic representations of my SqlA models and use those for the response_class and any Patch/Post data.

**Comment by ghost** [comment]
I'm seeing this when doing:

```
@dataclass
class Item:
    name: str

@dataclass
ItemResponse:
    item: List[Item]
```

The nested dataclass seems to break things. Otherwise, it was working for me. I wonder if the fieldfactory is what's breaking your example.

**Comment by ghost** [comment]
> Have you tried with a pydantic dataclass?
> 
> https://pydantic-docs.helpmanual.io/usage/dataclasses/

This works for me, but blows up in mypy.

I'll have to do:

```
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from dataclasses import dataclass
else:
    from pydantic.dataclasses import dataclass
```

Edit: Doesn't work well. Very broken. Responses, where I get them, have `__initialized__` or something inside dicts.

**Comment by whiskeyriver** [comment]
I think the common thread in all this is that Pydantic's BaseModel subclasses are the best suited type for the response.

**Comment by ghost** [comment]
Pydantic's BaseModels don't validate for me in mypy at all, which I find frustrating.

Edit, they do if there's no `= Field()` stuff going on. https://pydantic-docs.helpmanual.io/usage/mypy/

**Comment by Mause** [comment]
@sometato have you enabled the mypy plugin?

**Comment by ghost** [comment]
> @sometato have you enabled the mypy plugin?

I did, thank you. I'm not sure if it works with `Field()` that way or not, but it is working alright for me now.

I guess I like the feel and ease of dataclasses, but nested dataclasses break `/openapi.json`, and regardless they put `__initialized__` in every dictionary in the outputted JSON.

One thought might be to use `as_dict()` and dacite's `from_dict()` to go to/from dataclasses, instead of using Pydantic. Maybe could use `@app.get(dataclass=SomeDataclass)` instead of `response_model=`.

**Comment by alxdb** [comment]
I've been having this issue, and I think it is to do with the openapi generation functions encountering a native dataclass more than once. I can replicate the issue with this example:
```python
import dataclasses

import fastapi


@dataclasses.dataclass
class Device:
    id: int
    name: str
    

@dataclasses.dataclass
class User:
    main_device: Device
    other_device: Device


app = fastapi.FastAPI()


@app.post("/user")
def post_user(user: User):
    print(user)


app.openapi()
```

This results in `KeyError: <class 'pydantic.dataclasses.Device'>`

I've been looking through the fastapi source, and it seems that when a native dataclass is encountered, a pydantic dataclass is always generated for it _even if one has been generated previously_. This leads to issues that are to do with unexpected seemingly duplicate (although technically distinct) pydantic dataclasses.

I am currently working on a fix and will post a PR when ready.

EDIT:

This appears to be more of an issue on the pydantic side of the codebase, so I will open an issue there.

**Comment by stapetro** [comment]
Hello,

The change introduced #3576 broke the _fastapi.encoders.jsonable_encoder_ for Python native dataclasses. It doesn't produce JSON-compatible result when there's a dataclass containing a field of type Enum. Here's an example:
```python
import json
from dataclasses import dataclass
from enum import Enum

from fastapi.encoders import jsonable_encoder


class ResultState(Enum):
    OK = "Ok"
    ERROR = "Error"


@dataclass
class OpResult:
    state: ResultState


def test_fastapi_bug():
    op_res = OpResult(ResultState.OK)
    json_compatible_data = jsonable_encoder(op_res)

    json.dumps(json_compatible_data)

```
### Error
```bash
E       TypeError: Object of type ResultState is not JSON serializable
```
<details>
  <summary>Click to see full traceback</summary>
```bash
tests\unit\test_fastapi_bug.py:17 (test_fastapi_bug)
def test_fastapi_bug():
        op_res = OpResult(ResultState.OK)
        json_compatible_data = jsonable_encoder(op_res)
>       json.dumps(json_compatible_data)

unit\test_fastapi_bug.py:21: 
\_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
..\..\..\..\..\.pyenv\pyenv-win\versions\3.9.7\lib\json\__init__.py:231: in dumps
    return _default_encoder.encode(obj)
..\..\..\..\..\.pyenv\pyenv-win\versions\3.9.7\lib\json\encoder.py:199: in encode
    chunks = self.iterencode(o, _one_shot=True)
..\..\..\..\..\.pyenv\pyenv-win\versions\3.9.7\lib\json\encoder.py:257: in iterencode
    return _iterencode(o, 0)
\_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <json.encoder.JSONEncoder object at 0x000001AD72FB6100>
o = <ResultState.OK: 'Ok'>

    def default(self, o):
        """Implement this method in a subclass such that it returns
        a serializable object for ``o``, or calls the base implementation
        (to raise a ``TypeError``).
    
        For example, to support arbitrary iterators, you could
        implement default like this::
    
            def default(self, o):
                try:
                    iterable = iter(o)
                except TypeError:
                    pass
                else:
                    return list(iterable)
                # Let the base class default method raise the TypeError
                return JSONEncoder.default(self, o)
    
        """
\>       raise TypeError(f'Object of type {o.__class__.__name__} '
                        f'is not JSON serializable')
E       TypeError: Object of type ResultState is not JSON serializable

..\..\..\..\..\.pyenv\pyenv-win\versions\3.9.7\lib\json\encoder.py:179: TypeError
```
</details>

**Comment by radiantone** [comment]
You need to use the dataclass decorator from pydantic, not from dataclasses import dataclass.
Then it works.

**Comment by alxdb** [comment]
In the particular use case I had when I found this issue, I had to interface with existing standard lib dataclasses

**Comment by YuriiMotov** [comment, marked as answer]
Works fine in current version (fastapi 0.115.13, pydantic 1.10.22 or 2.11.7)
